pkg/database/sms_config.go

```diff
@@ -16,58 +16,113 @@ package database
+	TwilioAuthToken  string `gorm:"type:varchar(250)"`
 	TwilioFromNumber string `gorm:"type:varchar(16)"`
 
-	// This field contains the resolved secret, used for testing secrets were
-	// actually resolved properly.
-	twilioAuthSecret string
+	// twilioAuthToken and twilioAuthTokenEncrypted are cached values of the auth
+	// token. It's used to compare whether the value changed to avoid unnecessary
+	// calls the KMS.
+	twilioAuthToken          string `gorm:"-"`
+	twilioAuthTokenEncrypted string `gorm:"-"`
 }
 
-// GetSMSProvider gets the SMS provider for the given realm. The values are
-// cached.
-func (smsConfig *SMSConfig) GetSMSProvider(ctx context.Context, db *Database) (sms.Provider, error) {
-	// The provider itself is cached, not the SMSConfig entity.
-	key := fmt.Sprintf("GetSMSProvider/%v", smsConfig.ID)
-	val, err := db.cacher.WriteThruLookup(key, func() (interface{}, error) {
-		// Get the actual auth token from the secret manager.
-		//
-		// TODO(sethvargo): this is a double switch, since ProviderFor is also a
-		// switch. Maybe we should push this logic down lower? We only want to
-		// resolve secrets for the provider that's configured.
-		switch smsConfig.ProviderType {
-		case sms.ProviderTypeTwilio:
-			authToken, err := db.secretManager.GetSecretValue(ctx, smsConfig.TwilioAuthToken)
-			if err != nil {
-				return nil, fmt.Errorf("failed to resolve twilio auth token secret: %w", err)
-			}
-			smsConfig.twilioAuthSecret = authToken
+// BeforeSave runs before records are saved/created. It's used to mutate values
+// (such as the auth tokens) before storing them in the database. Do not call
+// this function, gorm calls it automatically.
+func (r *SMSConfig) BeforeSave() error {
+	ctx, done := context.WithTimeout(context.Background(), 5*time.Second)
+	defer done()
+
+	if plaintext := r.TwilioAuthToken; plaintext != "" {
+		// Don't encrypt again if the value hasn't changed.
+		if plaintext == r.twilioAuthToken {
+			r.TwilioAuthToken = r.twilioAuthTokenEncrypted
+			return nil
+		}
+
+		keyID := r.db.config.EncryptionKey
+		b, err := r.db.keyManager.Encrypt(ctx, keyID, []byte(plaintext), nil)
+		if err != nil {
+			return fmt.Errorf("failed to encrypt twilio auth token: %w", err)
+		}
+		r.TwilioAuthToken = base64.RawStdEncoding.EncodeToString(b)
+		r.twilioAuthToken = plaintext
+		r.twilioAuthTokenEncrypted = base64.RawStdEncoding.EncodeToString(b)
+	}
+	return nil
+}
+
+// AfterSave runs after records are saved. It's used to mutate values. Do not
+// call this function, gorm calls it automatically.
+func (r *SMSConfig) AfterSave() error {
+	if r.twilioAuthToken != "" {
+		r.TwilioAuthToken = r.twilioAuthToken
+	}
+	return nil
+}
+
+// AfterFind runs after a record is found. It's used to mutate values (such as
+// auth tokens) before storing them on the struct. Do not call this function,
+// gorm calls it automatically.
+func (r *SMSConfig) AfterFind() error {
```

Comment:
There needs to be some kind of auto upgrade from secret to encrypted in DB for existing installations.

Reply:
Added